holoviews/plotting/plot.py

```diff
@@ -597,6 +613,9 @@ class GenericElementPlot(DimensionedPlot):
     logy = param.Boolean(default=False, doc="""
         Whether the y-axis of the plot will be a log axis.""")
 
+    padding = param.ClassSelector(default=0, class_=(int, float, tuple), doc="""
+        Amount of padding to apply to data ranges.""")
```

Comment:
This still needs a clearer docstring; it's not clear what units it's in, whether it only applies to auto-ranged data or also to explicit ranges, etc.  Maybe:

```
Amount of extra screen space to add as a buffer around the specified or auto-ranged extents.
Specified as a fractional addition to width and heights, e.g., a padding of 0.2 will make an x,y plot 20% larger than the widest axis (x or y).  If specified as a tuple, the int or float values in the tuple will be used for padding in each axis, in order (x,y or x,y,z).
```

(not sure precisely how to describe it, but something like that?).

Reply:
Not even sure about the name ``padding`` yet, matplotlib uses the term ``margin`` while bokeh uses ``range_padding``.

> Amount of extra screen space to add as a buffer around the specified or auto-ranged extents.

This is not quite correct, no buffer is added to user specified values. Otherwise sounds fine.

Reply:
I don't think I like `margin`, as I expect that to be outside the plot, not in it.  Padding I don't have much of an expectation, and am more likely to look up what that means.  `range_padding` presumably isn't appropriate for calculations in screen space, so I prefer `padding`.  

My point was mainly that it's not clear whether it applies to specified values or not, which you've clarified.  (I could imagine an argument for applying it to user extents too, i.e. that you could automatically add a bit of space to user-computed values, but on balance I think it's probably least surprising not to.)  So it sounds like:

```
Amount of extra screen space to add as a buffer around auto-ranged extents.
Specified as a fractional addition to width and heights, e.g., a padding of 0.2 will make an x,y plot 20% larger than the widest axis (x or y).  If specified as a tuple, the int or float values in the tuple will be used for padding in each axis, in order (x,y or x,y,z).
```

If that's correct, then I think that's enough information.

Reply:
>Specified as a fractional addition to width and heights

Don't think this is quite the correct way of phrasing this, it sounds like it's actually adding to the width and height. Maybe something like

>Amount of extra screen space to add as a buffer around auto-ranged extents. Specified as a fraction of the width and height of an axis, ensuring equal padding for non-square plots. For example a padding of 0.2 will ensure that 20% of each axis range is devoted to padding. If specified as a tuple, the int or float values in the tuple will be used for padding in each axis, in order (x,y or x,y,z).

Reply:
I don't quite get it.  How will a value of 0.2 ensure that 20% of each axis range is devoted to padding, if we want the same padding all around, even for non-square axes?  Doesn't it need to be 20% of the axis range of the axis whose width or height is larger?

Reply:
>Doesn't it need to be 20% of the axis range of the axis whose width or height is larger?

You're right that's still correct, my objection was specifically to:

>Specified as a fractional addition to width and heights

But we do need to keep the bit about the larger axis.

Reply:
Ah, ok.  It's tough for me to write good docs when I'm not sure of the precise details of the algorithm, but that's why we need good docs!  I can't see how to say this concisely, so how about:

```
Fraction by which to increase auto-ranged extents to make datapoints more visible
around borders.

To compute padding, the axis whose screen size is largest is chosen, and the range
of that axis is increased by the specified fraction.  To ensure equal padding for all
axes, other axes are then padded so that their screen size will increase by roughly
the same amount.  If specified as a tuple, the int or float values in the tuple will be
used for padding in each axis, in order (x,y or x,y,z).

For example, for `padding=0.2` on a 800x800-pixel plot, an x axis with the range 
[0,10] will be padded by 20% to be [-1,11], while a y axis with a range [0,1000] will
be padded to be [-100,1100], which should make the padding be approximately the
same number of pixels.  But if the same plot is changed to have a height of only
200, the y range will then be [-400,1400] so that the y-axis padding will match that
of the x axis.
```

Reply:
>To ensure equal padding for all axes, other axes are then padded so that their screen size will increase by roughly the same amount.

This again makes it sound like the screen space is increased, when it simply devotes a certain percentage of the existing screen space to the padding. Your description walking through an example of what it's actually doing is correct though. How about changing that to

>Other axes are then padded ensuring that the amount of screen space devoted to padding is equal for all axes.

Reply:
>Fraction by which to increase auto-ranged extents to make datapoints more visible
around borders.

>To compute padding, the axis whose screen size is largest is chosen, and the range
of that axis is increased by the specified fraction.  Other axes are then padded ensuring that the amount of screen space devoted to padding is equal for all axes.  If specified as a tuple, the int or float values in the tuple will be used for padding in each axis, in order (x,y or x,y,z).

>For example, for `padding=0.2` on a 800x800-pixel plot, an x axis with the range 
[0,10] will be padded by 20% to be [-1,11], while a y axis with a range [0,1000] will
be padded to be [-100,1100], which should make the padding be approximately the
same number of pixels.  But if the same plot is changed to have a height of only
200, the y range will then be [-400,1400] so that the y-axis padding will match that
of the x axis.

Reply:
Good catch; yes, that's what I meant to convey, now much more clearly. I'm happy with both the description and the algorithm now, and look forward to seeing it in action!